Add MultiplayerSpawner interface to Multiplayer Gem

AMZN-puvvadar opened the pull request:
This change interfaces out the spawning of a player controlled autonomous prefab. Ownership of spawning a networked prefab now belongs to the user of the  Multiplayer Gem (i.e. MultiplayerSample). On join, the Multiplayer Gem now calls out to the interface expecting a NetworkEntityHandle back for it to mark as Autonomous. The interface also provides a hook for player disconnect to allow for cleanup.

In the event an IMultiplayerSpawner impl is not provided, an error is thrown indicating that an implementation is required.

Hooks into the Multiplayer Gem are tested via unit tests.

Signed-off-by: puvvadar <puvvadar@amazon.com>

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerSpawner.h:
```diff
@@ -0,0 +1,35 @@
+
+        virtual ~IMultiplayerSpawner() = default;
+
+        virtual AZStd::pair<Multiplayer::PrefabEntityId, AZ::Transform> SpawnPlayerPrefab(uint64_t userId) = 0;
```
Could this be a more generalized OnPlayerConnect that takes the userId and possibly other session parameters and returns the prefab/transform pair?

kberg-amzn replied:
Similarly, we should have an OnPlayerDisconnect that takes the disconnect reason, the autonomous entity list, etc... since different games want to handle disconnect differently (sometimes AI takes over, sometimes player despawns, sometimes it just dies and ragdolls, sometimes it will idle waiting on player reconnect, etc..)

AMZN-puvvadar replied:
A disconnect makes sense to me. Let me figure out how the despawn logic should look and I'll get this updated.

AMZN-puvvadar replied:
Okay disconnect is added. Did you have any particular thoughts on additional Connect params? I was thinking of adding a piece of config to the Connect packet we could pass through here but that was about it. Do you see a need for connection information?

kberg-amzn replied:
I was thinking to just pass across the MultiplayerAgentDatum declared in IMultiplayer.h.. presumably a game will handle identity verification through that 2k blob of data attached. Things like session tokens, join nonces, etc.. The customer game project server library could then open a connection to whatever service backend the customer is using, steam, epic, live, psn, gamelift, etc so the services could push information about pending player joins to validate against once the actual client connects start coming in.

kberg-amzn commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -515,7 +516,31 @@ namespace Multiplayer
         {
             // We use a temporary userId over the clients address so we can maintain client lookups even in the event of wifi handoff
-            NetworkEntityHandle controlledEntity = SpawnDefaultPlayerPrefab(packet.GetTemporaryUserId());
+            IMultiplayerSpawner* spawner = AZ::Interface<IMultiplayerSpawner>::Get();
```
We should make the filtered entity system and this system agree on how they register with the multiplayer gem just for API consistency. I don't really have a preference one way or another which style we end up picking, but try and work out with Olex which style you folks would like to move forward with.

AMZN-puvvadar replied:
I'll take a look through and come up with some options.

AMZN-Olex replied:
AZ::Interface::Get works as well, I assume that it won't affect perf versus a raw pointer but if it does, then we would need to know how much.

kberg-amzn commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -515,7 +516,31 @@ namespace Multiplayer
+            NetworkEntityHandle controlledEntity;
+
+            // Check rejoin data first
+            const auto node = m_playerRejoinData.find(packet.GetTemporaryUserId());
```
Hrm.. this was for multihost player migration.. the code here doesn't need to change since this is all negotiated at a higher level in a different GEM; but now that you're writing a formalized interface that would allow game specific rejoin behaviour I should probably rename some of these members to make things less confusing.

kberg-amzn commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -515,7 +516,31 @@ namespace Multiplayer
+            }
+            else if (!spawner)
+            {
+                // If no spawner we have a default option
```
Kill this codepath. I would rather assert than leave the cvar in place now that you're writing a real system.

kberg-amzn commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -853,7 +878,23 @@ namespace Multiplayer
+            {
+                // If no spawner we have a default option
+                // make sure the player prefab path is lowercase (how it's stored in the cache folder)
+                auto sv_defaultPlayerSpawnAssetLowerCase = static_cast<AZ::CVarFixedString>(sv_defaultPlayerSpawnAsset);
```
Same.. kill this cvar and codepath. We should *only* support the spawner interface system.

lmbr-pip reviewed (commented):
"This change interfaces out logic determining what entity is spawned and where when a user joins a session" maybe? "This change provides an interface for determining what entity is spawned and where, when a user joins a session"

AMZN-puvvadar commented:
Updates per last push:
- Added OnPlayerLeave to go with the now renamed OnPlayerJoin
- Removed default despawn logic in ServerToClientConnectionData
- Remove default spawn logic in MPSystemComponent
- Guarded for failure to spawn a controlled entity, I figure this is potentially desirable in the case of something like spectator mode

AMZN-puvvadar commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -84,8 +85,6 @@ namespace Multiplayer
     AZ_CVAR(bool, sv_isDedicated, true, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Whether the host command creates an independent or client hosted server");
     AZ_CVAR(bool, sv_isTransient, true, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Whether a dedicated server shuts down if all existing connections disconnect.");
     AZ_CVAR(AZ::TimeMs, sv_serverSendRateMs, AZ::TimeMs{ 50 }, nullptr, AZ::ConsoleFunctorFlags::Null, "Minimum number of milliseconds between each network update");
-    AZ_CVAR(AZ::CVarFixedString, sv_defaultPlayerSpawnAsset, "prefabs/player.network.spawnable", nullptr, AZ::ConsoleFunctorFlags::DontReplicate,
```
No, MultiplayerSample will. The main point is that this responsibility is moved to game code that uses the Multiplayer Gem. See https://github.com/o3de/o3de-multiplayersample/pull/106

AMZN-Gene replied:
My hesitation of not having a default is that it will add an extra step for devs wanting to get started with multiplayer (a process which already requires extra steps over other gems which are plug and play). 🤔

AMZN-puvvadar replied:
Again, see the MPSample pull request. The default is over there now.

AMZN-Gene replied:
The MPS Sample, is a perfect example. I'm saying it's _MultiplayerSample/Gem/Code/Source/MultiplayerSampleSystemComponent.h_
As opposed to _o3de/Gems/Multiplayer/Code/..._
It'll require most (all?) devs to do some copy+pasting after enabling the MultiplayerGem.
The Multiplayer Gem already requires some copy+pasting, but I notice ppl at first are confused by that since the other gems are plug and play.
One answer could be adding the Sample to the Multiplayer Template, although I think it would be cooler to have MP gem handle these sort of things automatically.

AMZN-Gene replied:
To put it another way, I'm having trouble imagining anyone not wanting to use your sweet spawner code when getting started :P

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerSpawner.h:
```diff
@@ -0,0 +1,37 @@
+    //! @brief IMultiplayerSpawner routes spawning requests for connecting players from
+    //! the Muliplayer Gem to game logic utilizing it
+    //!
+    //! IMultiplayerSpawner is an AZ::Interface<T> that provides a mechanism to
```
Agreed!

kberg-amzn commented on Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:
```diff
@@ -515,18 +514,36 @@ namespace Multiplayer
+                // Route to spawner implementation
+                AZStd::pair<PrefabEntityId, AZ::Transform> spawnParams = spawner->OnPlayerJoin(packet.GetTemporaryUserId());
+                controlledEntity = TrySpawnPlayerPrefab(spawnParams.first, spawnParams.second);
+            }
```
Pop out some kind of a visible error if you don't have a spawner. While I don't think we should support a cvar fallback path, we should make misconfiguration easy to detect and diagnose.

lmbr-pip reviewed (commented):
Echo Gene's concerns that we should have some very minimal code on the MP Gem side (if it makes sense) so theres some out-of-the-box experience or a clear indication for "hey you need to do work here".

MPS is a really useful tutorial space about how to do something but what are the user expectations those making a new multiplayer project? If they have to know they need to write code for spawning from day one, then we need to at least provide documentation of steps but ideally provide a highly simple but useful starting point.

AMZN-puvvadar commented:
Latest change addresses a spawning bug in which transform was not set in a particular spawning case making that information unavailable to dependent components at activation time. This fixes an issue where aim angles would blow out initial Z rotation.

kberg-amzn commented on Gems/Multiplayer/Code/Source/NetworkEntity/EntityReplication/EntityReplicator.cpp:
```diff
@@ -33,6 +33,29 @@
 {
+    void SyncActivatingEntityTransform(AZ::Entity* entity)
+    {
+        NetworkTransformComponent* netTransform = entity->FindComponent<NetworkTransformComponent>();
```
This find is likely necessary at the moment.

kberg-amzn commented on Gems/Multiplayer/Code/Source/NetworkEntity/EntityReplication/EntityReplicator.cpp:
```diff
@@ -33,6 +33,29 @@
+    void SyncActivatingEntityTransform(AZ::Entity* entity)
+    {
+        NetworkTransformComponent* netTransform = entity->FindComponent<NetworkTransformComponent>();
+        AzFramework::TransformComponent* transformComponent = entity->FindComponent<AzFramework::TransformComponent>();
```
But this find can be removed.. TransformComponent is cached since it's so frequently used. You can get the TransformInterface directly off the entity itself and avoid the linear component search.

AMZN-puvvadar replied:
Apparently this actually is needed. It's early enough that no TransformInterfaces are actually registered so the cache get fails. @SergeyAMZN might have a better understanding given he authored this piece.

SergeyAMZN replied:
Correct, it's because we sync **Activating** entity transform. Transform Component is cached **after** the entity is activated.
